src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -344,104 +354,223 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
+                                    string fullPath = Path.Combine(dirPath.FullPath, name);
+                                    Interop.libcoreclr.fileinfo fileinfo;
+                                    while (Interop.CheckIo(Interop.libcoreclr.GetFileInformationFromPath(fullPath, out fileinfo), fullPath)) ;
+                                    isDir = (fileinfo.mode & Interop.libcoreclr.FileTypes.S_IFMT) == Interop.libcoreclr.FileTypes.S_IFDIR;
+                                    isFile = (fileinfo.mode & Interop.libcoreclr.FileTypes.S_IFMT) == Interop.libcoreclr.FileTypes.S_IFREG;
+                                    break;
                             }
-                            if (searchOption == SearchOption.AllDirectories)
+                            bool matchesSearchPattern =
+                                (isFile || isDir) &&
+                                Interop.libc.fnmatch(_searchPattern, name, Interop.libc.FnmatchFlags.None) == 0;
+
+                            // Yield the result if the user has asked for it.  In the case of directories,
+                            // always explore it by pushing it onto the stack, regardless of whether
+                            // we're returning directories.
+                            if (isDir && !ShouldIgnoreDirectory(name))
                             {
-                                if (toExplore == null)
+                                if (_includeDirectories && matchesSearchPattern)
+                                {
+                                    yield return _translateResult(Path.Combine(dirPath.UserPath, name), /*isDirectory*/true);
+                                }
+                                if (_searchOption == SearchOption.AllDirectories)
                                 {
-                                    toExplore = new Stack<string>();
+                                    if (toExplore == null)
+                                    {
+                                        toExplore = new Stack<PathPair>();
+                                    }
+                                    toExplore.Push(new PathPair(Path.Combine(dirPath.UserPath, name), Path.Combine(dirPath.FullPath, name)));
                                 }
-                                toExplore.Push(fullNewName);
+                            }
+                            else if (isFile && _includeFiles && matchesSearchPattern)
+                            {
+                                yield return _translateResult(Path.Combine(dirPath.UserPath, name), /*isDirectory*/false);
                             }
                         }
-                        else if (isFile && includeFiles && matchesSearchPattern)
+                    }
+                    finally
+                    {
+                        // Close the directory enumerator
+                        if (dirHandle != null && !dirHandle.IsClosed) // TODO: Remove once exception handling bug is fixed
```

Comment:
I'll remove this before checking in once https://github.com/dotnet/coreclr/pull/674 is merged.